lib/SILOptimizer/Transforms/SimplifyCFG.cpp

```diff
@@ -916,16 +916,44 @@ static bool couldRemoveRelease(SILBasicBlock *SrcBB, SILValue SrcV,
   return IsReleaseOfDest;
 }
 
+/// Returns true if any instruction in \p block may write memory.
+static bool blockMayWriteMemory(SILBasicBlock *block) {
+  return std::any_of(block->begin(), block->end(), [](SILInstruction &inst) {
```

Comment:
I think this should just bail when the number of instructions is some multiple of the expected increase in the threading budget. It's the pathological case of several thousand read-only instruction in row that matters for compile time.